README.md
# pmsampsize <img src="man/figures/logo.png" align="right" height="139" alt="" />

<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/pmsampsize)](https://CRAN.R-project.org/package=pmsampsize)
[![CRAN downloads](https://cranlogs.r-pkg.org/badges/grand-total/pmsampsize)](https://CRAN.R-project.org/package=pmsampsize)
[![R-CMD-check](https://github.com/joieensor/pmsampsize/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/joieensor/pmsampsize/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

> **Note:** Replace `joieensor` in the badge URLs above with your actual GitHub username.

## Overview

`pmsampsize` computes the **minimum sample size required for the development of a new multivariable prediction model**, using the criteria proposed by Riley *et al.* (2018, 2020). It supports models with:

- **Continuous** outcomes (linear prediction models)
- **Binary** outcomes (logistic prediction models)
- **Survival / time-to-event** outcomes (Cox prediction models)

The criteria implemented aim to minimise overfitting and ensure precise estimation of key parameters in the prediction model.

## Installation

Install the released version from CRAN:

```r
install.packages("pmsampsize")
```

Or install the development version from GitHub:

```r
# install.packages("remotes")
remotes::install_github("joieensor/pmsampsize")
```

> **Note:** Replace `joieensor` with your actual GitHub username.

## Usage

### Binary outcome

Calculate the minimum sample size for a logistic prediction model with 24 candidate predictor parameters, an anticipated outcome prevalence of 17.4%, and a Cox-Snell R² of 0.288 (taken from an existing model):

```r
library(pmsampsize)

pmsampsize(type = "b", csrsquared = 0.288, parameters = 24, prevalence = 0.174)
```

If only a C-statistic (rather than R²) is available from an existing model, use the `cstatistic` argument instead:

```r
pmsampsize(type = "b", cstatistic = 0.89, parameters = 24, prevalence = 0.174)
```

### Survival outcome

Calculate the minimum sample size for a Cox prediction model with 30 candidate predictors, using information from an existing study (R² = 0.051, mean follow-up = 2.07 years, event rate = 0.065), with a prediction timepoint of 2 years:

```r
pmsampsize(type = "s", csrsquared = 0.051, parameters = 30, rate = 0.065,
           timepoint = 2, meanfup = 2.07)
```

### Continuous outcome

Calculate the minimum sample size for a linear prediction model with 25 candidate predictors, R² = 0.2, mean outcome value of 1.9, and SD of 0.6:

```r
pmsampsize(type = "c", rsquared = 0.2, parameters = 25, intercept = 1.9, sd = 0.6)
```

## Criteria implemented

### Continuous outcomes — four criteria:
1. Small overfitting: expected shrinkage of predictor effects ≤ 10%
2. Small absolute difference (≤ 0.05) between apparent and adjusted R²
3. Precise estimation of the residual standard deviation
4. Precise estimation of the average outcome value

### Binary and survival outcomes — three criteria:
1. Small overfitting: expected shrinkage of predictor effects ≤ 10%
2. Small absolute difference (≤ 0.05) between apparent and adjusted Nagelkerke's R²
3. Precise estimation (within ± 0.05) of the average outcome risk at the timepoint of interest

## Key arguments

| Argument | Type(s) | Description |
|---|---|---|
| `type` | all | `"c"` continuous, `"b"` binary, `"s"` survival |
| `parameters` | all | Number of candidate predictor parameters |
| `shrinkage` | all | Target shrinkage (default = 0.9) |
| `rsquared` | `"c"` | Anticipated R² of new model |
| `csrsquared` | `"b"`, `"s"` | Anticipated Cox-Snell R² |
| `nagrsquared` | `"b"`, `"s"` | Anticipated Nagelkerke's R² |
| `cstatistic` | `"b"` | C-statistic from existing model (used to approximate Cox-Snell R²) |
| `prevalence` | `"b"` | Anticipated outcome proportion |
| `rate` | `"s"` | Overall event rate in the population |
| `timepoint` | `"s"` | Timepoint of interest for prediction |
| `meanfup` | `"s"` | Anticipated mean follow-up time |
| `intercept` | `"c"` | Average outcome value in the population |
| `sd` | `"c"` | Standard deviation of outcome values |
| `mmoe` | `"c"` | Multiplicative margin of error for intercept (default = 1.1) |

## Citation

If you use `pmsampsize` in your research, please cite the following papers:

Riley RD, Ensor J, Snell KIE, Harrell FE, Martin GP, Reitsma JB, et al. Calculating the sample size required for developing a clinical prediction model. *BMJ*. 2020. <https://doi.org/10.1136/bmj.m441>

Riley RD, Snell KIE, Ensor J, Burke DL, Harrell FE Jr., Moons KG, Collins GS. Minimum sample size required for developing a multivariable prediction model: Part I — continuous outcomes. *Statistics in Medicine*. 2018. <https://doi.org/10.1002/sim.7993>

Riley RD, Snell KIE, Ensor J, Burke DL, Harrell FE Jr., Moons KG, Collins GS. Minimum sample size required for developing a multivariable prediction model: Part II — binary and time-to-event outcomes. *Statistics in Medicine*. 2018. <https://doi.org/10.1002/sim.7992>

Riley RD, Van Calster B, Collins GS. A note on estimating the Cox-Snell R² from a reported C statistic (AUROC) to inform sample size calculations for developing a prediction model with a binary outcome. *Statistics in Medicine*. 2020.

## Author

**Joie Ensor** (University of Birmingham, [j.ensor@bham.ac.uk](mailto:j.ensor@bham.ac.uk))

With thanks to Richard D. Riley, Emma C. Martin, Gary Collins, Glen Martin & Kym Snell for helpful input and feedback.

## License

GPL (≥ 3)


NEWS.md
pmsampsize v1.1.3 (Release date: 2023-12-05)
=============================================
* added options to specify the cox-snell or nagelkerke's r-squared directly for binary outcome models
* simplified continuous outcome models criteria 3 calculation fixing MMOE=10% so that N=234+p

pmsampsize v1.1.2 (Release date: 2022-02-12)
=============================================
* fixes issue for continuous outcome models in criteria 4 in calculating the CI for the intercept based on final SS
* edit to criteria 2 for continuous outcomes to round up N

pmsampsize v1.1.1 (Release date: 2021-11-21)
=============================================
* fixes issues with rounding leading to small differences in Max_Rsq and subsequently sample size calculations
* additional output including Nagelkerke's R-sq (binary & survival outcomes)

pmsampsize v1.1.0 (Release date: 2021-06-24)
=============================================
* update allowing user to input C-statistic from existing model to approximate Cox-Snell R-squared (for binary outcome model sample size calculations) 
* additional parameter checks	

pmsampsize v1.0.4 (Release date: 2020-07-24)
=============================================
* version issues fixes

pmsampsize v1.0.3 (Release date: 2020-07-23)
=============================================

Continuous outcomes:

* calculation of shrinkage for each criteria based on N needed in specific criteria

* also allowance for the unique scenario when starting N already meets shrinkage requirement set by user (this is more for academic exercise, as in practice a high shrinkage will be desired and so it is very rare that the base N (set to p+2) will provide good shrinkage)

Binary and TTE outcomes:

* more informative error message added for rare scenario where shrinkage is specified lower than r-squared which results in NaNs from shrinkage formula 

pmsampsize v1.0.2 (Release date: 2019-11-15)
=============================================

* update forces the time-unit to be the mean follow-up time, as described in forthcoming BMJ paper which extends the original Statistics in Medicine paper (see reference in help file)

* updated in calculation of max R-sq for survival option, to prevent manipulation based on time units used 


pmsampsize v1.0.1 (Release date: 2019-08-02)
=============================================

* updated in line with correction to original article in eq. 13


pmsampsize v1.0.0 (Release date: 2019-01-02)
=============================================

* this is a new release



tests/testthat/test_bin.R
context("binary_ss_checks")

test_that("Test 4.1", {
  ss <- c(623, 662, 221, 662)
  SPP <- c(4.52, 4.8, 1.6, 4.8)
  test <- pmsampsize(type="b", csrsquared= 0.288, parameters=24, prevalence = 0.174)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,7]
  names(test_ss) <- names(test_SPP) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
})

test_that("Test 4.2", {
  ss <- c(255, 532, 221, 532)
  SPP <- c(1.85, 3.86, 1.6, 3.86)
  test <- pmsampsize(type="b", csrsquared= 0.55, parameters=24, prevalence = 0.174)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,7]
  names(test_ss) <- names(test_SPP) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
})

test_that("Test 4.3", {
  ss <- c(934, 992, 221, 992)
  SPP <- c(4.51, 4.79, 1.07, 4.79)
  test <- pmsampsize(type="b", csrsquared= 0.288, parameters=36, prevalence = 0.174)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,7]
  names(test_ss) <- names(test_SPP) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
})

test_that("Test 4.4", {
  ss <- c(623, 590, 289, 623)
  SPP <- c(6.49, 6.15, 3.01, 6.49)
  test <- pmsampsize(type="b", csrsquared= 0.288, parameters=24, prevalence = 0.25)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,7]
  names(test_ss) <- names(test_SPP) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
})




tests/testthat/test_cont.R
context("continuous_ss_checks")

test_that("Test 2.1", {
  ss <- c(918, 401, 259, 918, 918)
  SPP <- c(36.72, 16.04, 10.36, 36.72, 36.72)
  test <- pmsampsize(type="c",rsquared=0.2,parameters=25,intercept=1.9,sd=.6)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,5]
  names(test_ss) <- names(test_SPP) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
})

test_that("Test 2.2", {
  ss <- c(574, 351, 259, 574, 574)
  SPP <- c(22.96, 14.04, 10.36, 22.96, 22.96)
  test <- pmsampsize(type="c",rsquared=0.3,parameters=25,intercept=1.9,sd=.6)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,5]
  names(test_ss) <- names(test_SPP) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
})

test_that("Test 2.3", {
  ss <- c(1924, 801, 284, 1924, 1924)
  SPP <- c(38.48, 16.02, 5.68, 38.48, 38.48)
  test <- pmsampsize(type="c",rsquared=0.2,parameters=50,intercept=1.9,sd=.6)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,5]
  names(test_ss) <- names(test_SPP) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
})

test_that("Test 2.4", {
  ss <- c(918, 401, 259, 918, 918)
  SPP <- c(36.72, 16.04, 10.36, 36.72, 36.72)
  test <- pmsampsize(type="c",rsquared=0.2,parameters=25,intercept=10,sd=.6)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,5]
  names(test_ss) <- names(test_SPP) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
})

test_that("Test 2.5", {
  ss <- c(918, 401, 259, 918, 918)
  SPP <- c(36.72, 16.04, 10.36, 36.72, 36.72)
  test <- pmsampsize(type="c",rsquared=0.2,parameters=25,intercept=1.9,sd=2)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,5]
  names(test_ss) <- names(test_SPP) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
})

test_that("Test 2.6", {
  ss <- c(1264, 1001, 334, 1264, 1264)
  SPP <- c(12.64, 10.01, 3.34, 12.64, 12.64)
  test <- pmsampsize(type="c",rsquared=0.5,parameters=100,intercept=1.9,sd=2)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,5]
  names(test_ss) <- names(test_SPP) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
})

test_that("Test 2.7", {
  ss <- c(32, 67, 264, 264, 264)
  SPP <- c(1.07, 2.23, 8.8, 8.8, 8.8)
  shrink <- c(0.845, 0.851, 0.954, 0.954, 0.954)
  test <- pmsampsize(type="c",rsquared=0.89,parameters=30,intercept=3.3,sd=1.1,shrinkage = 0.84)
  test_ss <- test$results_table[,1]
  test_SPP <- test$results_table[,5]
  test_shrink <- test$results_table[,2]
  names(test_ss) <- names(test_SPP) <- names(test_shrink) <- c()
  expect_equal(test_ss,ss)
  expect_equal(test_SPP,SPP)
  expect_equal(test_shrink,shrink)
})


R/cstat2rsq.R

cstat2rsq <- function(cstatistic, prevalence, seed = 123456){

  # define ss for simulation dataset
  n <- 1000000

  # define mu as a function of the C-statistic
  mu   <- sqrt(2) * qnorm(cstatistic)

  # simulate large sample linear prediction based on two normals
  # for non-eventsN(0, 1), events and N(mu, 1)
  set.seed(seed)
  LP <- c(rnorm(prevalence*n,  mean=0, sd=1), rnorm((1-prevalence)*n, mean=mu, sd=1))
  y <- c(rep(0, prevalence*n), rep(1, (1-prevalence)*n))

  # Fit a logistic regression with LP as covariate;
  # this is essentially a calibration model, and the intercept and
  # slope estimate will ensure the outcome proportion is accounted
  # for, without changing C-statistic
  model <- glm( y ~LP, family = binomial)

  return(list(R2.coxsnell = as.numeric(1-exp(-1*(model$null.deviance-model$deviance)/n))
              ))
}


#q <- cstat2rsq(cstatistic = 0.81, prevalence = 0.77)


R/pmsampsize.R
#' pmsampsize
#' - Sample Size for Development of a Prediction Model
#'
#' @description \code{pmsampsize} computes the minimum sample size required for the development of a new
#' multivariable prediction model using the criteria proposed by Riley \emph{et al}. 2018.
#'
#' @details \code{pmsampsize} can be used to calculate the minimum sample size for the development of models with
#' continuous, binary or survival (time-to-event) outcomes. Riley \emph{et al}. lay out a series of
#' criteria the sample size should meet. These aim to minimise the overfitting and to ensure
#' precise estimation of key parameters in the prediction model. \cr \cr
#' For continuous outcomes, there are four criteria: \cr
#' i) small overfitting defined by an expected shrinkage of predictor effects by 10\% or less, \cr
#' ii) small absolute difference of 0.05 in the model's apparent and adjusted R-squared value, \cr
#' iii) precise estimation of the residual standard deviation, and \cr
#' iv) precise estimation of the average outcome value. \cr \cr
#' The sample size calculation requires the user to pre-specify (e.g. based on previous evidence)
#' the anticipated R-squared of the model, and the average outcome value and standard deviation
#' of outcome values in the population of interest. \cr \cr
#' For binary or survival (time-to-event) outcomes, there are three criteria: \cr
#' i) small overfitting defined by an expected shrinkage of predictor effects by 10\% or less, \cr
#' ii) small absolute difference of 0.05 in the model's apparent and adjusted Nagelkerke's R-squared
#' value, and \cr
#' iii) precise estimation (within +/- 0.05) of the average outcome risk in the
#' population for a key timepoint of interest for prediction.
#'
#'
#' With thanks to Richard D. Riley, Emma C Martin, Gary Collins, Glen Martin & Kym Snell for helpful input & feedback
#'
#' @return A list including a matrix of calculated sample size requirements for each criteria defined
#' under \emph{'Details'}, and a series of vectors of parameters used in the calculations as well as the
#' final recommended minimum sample size and number of events required for model development.
#'
#' @author Joie Ensor (University of Birmingham, j.ensor@bham.ac.uk),
#'
#' @param type specifies the type of analysis for which sample size is being calculated
#'      \itemize{
#'          \item \code{"c"} specifies sample size calculation for a prediction model with a continuous outcome
#'          \item \code{"b"} specifies sample size calculation for a prediction model with a binary outcome
#'          \item \code{"s"} specifies sample size calculation for a prediction model with a survival (time-to-event) outcome
#'      }
#'
#' @param rsquared for \code{type="c"} this specifies the expected value of the R-squared of the new model,
#' where R-squared is the percentage of variation in outcome values explained by the model.
#' For example, the user may input the value of the R-squared reported for a
#' previous prediction model study in the same field.  If taking a value from a previous
#' prediction model development study, users should input the model's adjusted R-squared
#' value, not the apparent R-squared value, as the latter is optimistic (biased).  However,
#' if taking the R-squared value from an external validation of a previous model, the
#' apparent R-squared can be used (as the validation data was not used for development, and
#' so R-squared apparent is then unbiased).
#' Users should be conservative with their chosen R-squared value; for example, by taking the R-squared
#' value from a previous model, even if they hope their new model will improve performance.
#'
#' @param csrsquared for \code{type="b" or type="s"} this specifies the expected value of
#' the Cox-Snell R-squared of the new model. The Cox-Snell R-squared is the generalised
#' version of the well-known R-squared for continuous outcomes, based on the likelihood.
#' Please read the description of \code{rsquared} for additional details about specifying the
#' expected R-squared performance.
#' The papers by Riley et al. (see references) outline how to obtain the Cox-Snell R-squared value
#' from published studies if they are not reported, using other information (such as the
#' C-statistic [see \code{cstatistic()} option below]).
#'
#' @param nagrsquared for \code{type="b" or type="s"} this specifies the expected value of
#' the Nagelkerke's R-squared of the new model, which is the Cox-Snell R-squared scaled to
#' lie in the [0,1] range. It is interpretable in the same way as the standard R-squared, i.e.
#' the percentage of variation in outcome values explained by the model.
#' Please read the description of \code{rsquared} for additional details about specifying the
#' expected R-squared performance
#'
#' @param parameters specifies the number of candidate predictor parameters for potential
#' inclusion in the new prediction model.  Note that this may be larger than the number of
#' candidate predictors, as categorical and continuous predictors often require two or more
#' parameters to be estimated.
#'
#' @param shrinkage specifies the level of shrinkage desired at internal validation after
#' developing the new model. Shrinkage is a measure of overfitting, and can range from 0 to 1,
#' with higher values denoting less overfitting. We recommend a shrinkage = 0.9 (the
#' default in \code{pmsampsize}), which indicates that the predictor effect (beta coefficients) in
#' the model would need to be shrunk by 10\% to adjust for overfitting. See references
#' below for further information.
#'
#' @param prevalence (\code{type="b"} option) specifies the overall outcome proportion
#' (for a prognostic model) or
#' overall prevalence (for a diagnostic model) expected within the model development
#' dataset. This should be derived based on previous studies in the same population.
#'
#' @param cstatistic (\code{type="b"} option) specifies the C-statistic reported in an
#' existing prediction model study to be used in conjunction with the expected
#' prevalence to approximate the Cox-Snell R-squared using the approach of Riley et al. 2020.
#' Ideally, this should be an optimism-adjusted C-statistic. The approximate Cox-Snell R-squared
#' value is used as described above for the \code{csrsquared()} option, and so is treated as a baseline
#' for the expected performance of the new model.
#'
#' @param seed (\code{type="b"} option) specifies the initial value of the random-number
#' seed used by the random-number functions when simulating data to approximate the
#' Cox-Snell R-squared based on reported C-statistic and expect prevalence as described
#' by Riley et al. 2020
#'
#' @param rate (\code{type="s"} option) specifies the overall event rate in the population of interest,
#' for example as obtained from a previous study, for the survival outcome of interest. NB: rate must
#' be given in time units used for meanfup and timepoint options.
#'
#' @param timepoint (\code{type="s"} option) specifies the timepoint of interest for prediction.
#' NB: time units must be the same as given for meanfup option (e.g. years, months).
#'
#' @param meanfup (\code{type="s"} option) specifies the average (mean) follow-up time
#' anticipated for individuals in the model development dataset,
#' for example as taken from a previous study in the population of interest.
#' NB: time units must be the same as given for timepoint option.
#'
#' @param  intercept (\code{type="c"} options) specifies the average outcome value in the population of
#' interest e.g. the average blood pressure, or average pain score.
#' This could be based on a previous study, or on clinical knowledge.
#'
#' @param sd (\code{type="c"} options) specifies the standard deviation (SD) of
#' outcome values in the population e.g.
#' the SD for blood pressure in patients with all other predictors set to the average.
#' This could again be based on a previous study, or on clinical knowledge.
#'
#' @param mmoe (\code{type="c"} options) multiplicative margin of error (MMOE)
#' acceptable for calculation of the
#' intercept. The default is a MMOE of 10\%. Confidence interval for the intercept will be
#' displayed in the output for reference. See references below for further information.
#'
#' @import stats
#'
#' @examples
#' ## Examples based on those included in two papers by Riley et al.
#' ## published in Statistics in Medicine (2018).
#' ## NB: Survival example based on Riley et al. BMJ paper (2020).
#'
#' ## Binary outcomes (Logistic prediction models)
#' # Use pmsampsize to calculate the minimum sample size required to develop a
#' # multivariable prediction model for a binary outcome using 24 candidate
#' # predictor parameters. Based on previous evidence, the outcome prevalence is
#' # anticipated to be 0.174 (17.4%) and a lower bound (taken from the adjusted
#' # Cox-Snell R-squared of an existing prediction model) for the new model's
#' # R-squared value is 0.288
#'
#' pmsampsize(type = "b", csrsquared = 0.288, parameters = 24, prevalence = 0.174)
#'
#' # Now lets assume we could not obtain a Cox-Snell R-squared estimate from an existing
#' # prediction model, but instead had a C-statistic (0.89) reported for the existing prediction
#' # model. We can use this C-statistic along with the prevalence to approximate the Cox-Snell
#' # R-squared using the approach of Riley et al. (2020). Use pmsampsize with the cstatistic()
#' # option instead of rsquared() option.
#'
#' pmsampsize(type = "b", cstatistic = 0.89, parameters = 24, prevalence = 0.174)
#'
#' ## Survial outcomes (Cox prediction models)
#' # Use pmsampsize to calculate the minimum sample size required for developing
#' # a multivariable prediction model with a survival outcome using 30 candidate
#' # predictors. We know an existing prediction model in the same field has an
#' # R-squared adjusted of 0.051. Further, in the previous study the mean
#' # follow-up was 2.07 years, and overall event rate was 0.065. We select a
#' # timepoint of interest for prediction using the newly developed model of 2
#' # years
#'
#' pmsampsize(type = "s", csrsquared = 0.051, parameters = 30, rate = 0.065,
#'            timepoint = 2, meanfup = 2.07)
#'
#' ## Continuous outcomes (Linear prediction models)
#' # Use pmsampsize to calculate the minimum sample size required for developing
#' # a multivariable prediction model for a continuous outcome (here, FEV1 say),
#' # using 25 candidate predictors.  We know an existing prediction model in the
#' # same field has an R-squared adjusted of 0.2, and that FEV1 values in the
#' # population have a mean of 1.9 and SD of 0.6
#'
#' pmsampsize(type = "c", rsquared = 0.2, parameters = 25, intercept = 1.9, sd = 0.6)
#'
#' @references  Riley RD, Ensor J, Snell KIE, Harrell FE, Martin GP, Reitsma JB, et al.
#' Calculating the sample size required for developing a clinical prediction model.
#' \emph{BMJ (Clinical research ed)}. 2020
#'
#' @references  Riley RD, Snell KIE, Ensor J, Burke DL, Harrell FE, Jr., Moons KG, Collins GS.
#' Minimum sample size required for developing a multivariable prediction model: Part I continuous outcomes.
#' \emph{Statistics in Medicine}. 2018 (in-press). doi: 10.1002/sim.7993
#'
#' @references Riley RD, Snell KIE, Ensor J, Burke DL, Harrell FE, Jr., Moons KG, Collins GS.
#' Minimum sample size required for developing a multivariable prediction model:
#' Part II binary and time-to-event outcomes.
#' \emph{Statistics in Medicine}. 2018 (in-press). doi: 10.1002/sim.7992
#'
#' @references Riley, RD, Van Calster, B, Collins, GS. A note on estimating the Cox-Snell R2
#' from a reported C statistic (AUROC) to inform sample size calculations for developing a
#' prediction model with a binary outcome. \emph{Statistics in Medicine}. 2020


#' @export
pmsampsize <- function(type,
                       nagrsquared = NA,
                       csrsquared = NA,
                       rsquared = NA,
                       parameters,
                       shrinkage = 0.9,
                       prevalence = NA,
                       cstatistic = NA,
                       seed = 123456,
                       rate = NA,
                       timepoint = NA,
                       meanfup = NA,
                       intercept = NA,
                       sd = NA,
                       mmoe=1.1) {

  # error checking
  pmsampsize_errorcheck(type=type,nagrsquared=nagrsquared,csrsquared=csrsquared,rsquared=rsquared,parameters=parameters,shrinkage=shrinkage,cstatistic=cstatistic,
            prevalence=prevalence,rate=rate,timepoint=timepoint,meanfup=meanfup,intercept=intercept,sd=sd,mmoe=mmoe)

  # choose function based on analysis type
  if (type == "c") out <- pmsampsize_cont(rsquared=rsquared,parameters=parameters,intercept=intercept,
                                   sd=sd,shrinkage=shrinkage,mmoe=mmoe)

  if (type == "b") {
    rsquared <- NA
    if (is.na(nagrsquared) == F) {

      if (is.na(csrsquared) == F) {
        stop("Only one of csrsquared() or nagrsquared() can be specified")
      }

       E <- parameters*prevalence
			 lnLnull <- (E*(log(E/parameters)))+((parameters-E)*(log(1-(E/parameters))))
       max_r2a <- (1- exp((2*lnLnull)/parameters))
			 rsquared <- round(nagrsquared*max_r2a,digits=3)

    }

    if (is.na(csrsquared) == F) {
      if (is.na(nagrsquared) == F) {
        stop("Only one of csrsquared() or nagrsquared() can be specified")
      }
      rsquared <- csrsquared
    }

    if (is.na(cstatistic) == F) {

      if (is.na(rsquared) == F) {
        stop("Only one of csrsquared() or nagrsquared() or cstatistic() can be specified")
      }

      approx_rsq <- cstat2rsq(cstatistic=cstatistic, prevalence=prevalence, seed=seed)
      rsquared <- round(approx_rsq$R2.coxsnell,digits = 4)
    }
    else {
      if (is.na(rsquared)) {
        stop("One of csrsquared() or nagrsquared() or cstatistic() must be specified")
      }
    }
    out <- pmsampsize_bin(rsquared=rsquared,parameters=parameters,prevalence=prevalence,
                                         shrinkage=shrinkage,cstatistic=cstatistic)
  }

  if (type == "s") {
    rsquared <- NA
    if (is.na(nagrsquared) == F) {

      if (is.na(csrsquared) == F) {
        stop("Only one of csrsquared() or nagrsquared() can be specified")
      }

       events <- parameters*rate*meanfup
       lnLnull <- (events*(log(events/parameters)))-events
		 max_r2a <- (1- exp((2*lnLnull)/parameters))
      rsquared <- round(nagrsquared*max_r2a,digits=3)

    }

    if (is.na(csrsquared) == F) {
      if (is.na(nagrsquared) == F) {
        stop("Only one of csrsquared() or nagrsquared() can be specified")
      }
      rsquared <- csrsquared
    }
    else {
      if (is.na(rsquared)) {
        stop("One of csrsquared() or nagrsquared() must be specified")
      }
    }

    out <- pmsampsize_surv(rsquared=rsquared,parameters=parameters,rate=rate,
                                          timepoint=timepoint,meanfup=meanfup,shrinkage=shrinkage,mmoe=mmoe)
  }


  est <- out
  class(est) <- "pmsampsize"
  est


}


#' @export
print.pmsampsize <- function(x, ...) {

  if (x$type == "continuous") {
    cat("NB: Assuming 0.05 acceptable difference in apparent & adjusted R-squared \n")
    cat("NB: Assuming MMOE <= 1.1 in estimation of intercept & residual standard deviation \n")
    cat("SPP - Subjects per Predictor Parameter \n \n")

  }
  if (x$type == "survival") {
    cat("NB: Assuming 0.05 acceptable difference in apparent & adjusted R-squared \n")
    cat("NB: Assuming 0.05 margin of error in estimation of overall risk at time point =", x$timepoint, " \n")
    cat("NB: Events per Predictor Parameter (EPP) assumes overall event rate =", x$rate, " \n \n")
  }
  if (x$type == "binary") {
    if (is.na(x$cstatistic) == F) {
      cat("Given input C-statistic =", x$cstatistic, " & prevalence =", x$prevalence, " \n")
      cat("Cox-Snell R-sq =", x$rsquared, " \n \n")
    }
    cat("NB: Assuming 0.05 acceptable difference in apparent & adjusted R-squared \n")
    cat("NB: Assuming 0.05 margin of error in estimation of intercept \n")
    cat("NB: Events per Predictor Parameter (EPP) assumes prevalence =", x$prevalence, " \n \n")
  }
  print(x$results_table)

  if (x$type == "continuous") {
    cat(" \n Minimum sample size required for new model development based on user inputs =", x$sample_size, " \n \n ")
    cat("* 95% CI for intercept = (",x$int_lci,", ", x$int_uci, "), for sample size n = ",x$sample_size,sep = "")
  }
  if (x$type == "survival") {
    cat(" \n Minimum sample size required for new model development based on user inputs = ", x$sample_size, ", \n ",sep = "")
    cat("corresponding to", x$tot_per_yrs_final, "person-time** of follow-up, with", ceiling(x$events), "outcome events \n ")
    cat("assuming an overall event rate =", x$rate, "and therefore an EPP =", x$EPP, " \n \n ")
    cat("* 95% CI for overall risk = (",x$int_lci,", ", x$int_uci, "), for true value of ", x$int_cuminc, " and sample size n = ",x$sample_size,sep = "")
    cat(" \n **where time is in the units mean follow-up time was specified in")
  }
  if (x$type == "binary") {
    cat(" \n Minimum sample size required for new model development based on user inputs = ", x$sample_size, ", \n ",sep = "")
    cat("with ", ceiling(x$events), " events (assuming an outcome prevalence = ", x$prevalence, ") and an EPP = ", x$EPP, " \n \n ",sep = "")

  }
}

#' @export
summary.pmsampsize <- function(object, ...) {

  if (object$type == "continuous") {
    cat("NB: Assuming 0.05 acceptable difference in apparent & adjusted R-squared \n")
    cat("NB: Assuming MMOE <= 1.1 in estimation of intercept & residual standard deviation \n")
    cat("SPP - Subjects per Predictor Parameter \n \n")

  }
  if (object$type == "survival") {
    cat("NB: Assuming 0.05 acceptable difference in apparent & adjusted R-squared \n")
    cat("NB: Assuming 0.05 margin of error in estimation of overall risk at time point =", object$timepoint, " \n")
    cat("NB: Events per Predictor Parameter (EPP) assumes overall event rate =", object$rate, " \n \n")
  }
  if (object$type == "binary") {
    if (is.na(object$cstatistic) == F) {
      cat("Given input C-statistic =", object$cstatistic, " & prevalence =", object$prevalence, " \n")
      cat("Cox-Snell R-sq =", object$rsquared, " \n \n")
    }
    cat("NB: Assuming 0.05 acceptable difference in apparent & adjusted R-squared \n")
    cat("NB: Assuming 0.05 margin of error in estimation of intercept \n")
    cat("NB: Events per Predictor Parameter (EPP) assumes prevalence =", object$prevalence, " \n \n")
  }
  print(object$results_table)

  if (object$type == "continuous") {
    cat(" \n Minimum sample size required for new model development based on user inputs =", object$sample_size, " \n \n ")
    cat("* 95% CI for intercept = (",object$int_lci,", ", object$int_uci, "), for sample size n = ",object$sample_size,sep = "")
  }
  if (object$type == "survival") {
    cat(" \n Minimum sample size required for new model development based on user inputs = ", object$sample_size, ", \n ",sep = "")
    cat("corresponding to", object$tot_per_yrs_final, "person-time** of follow-up, with", ceiling(object$events), "outcome events \n ")
    cat("assuming an overall event rate =", object$rate, "and therefore an EPP =", object$EPP, " \n \n ")
    cat("* 95% CI for overall risk = (",object$int_lci,", ", object$int_uci, "), for true value of ", object$int_cuminc, " and sample size n = ",object$sample_size,sep = "")
    cat(" \n **where time is in the units mean follow-up time was specified in")
  }
  if (object$type == "binary") {
    cat(" \n Minimum sample size required for new model development based on user inputs = ", object$sample_size, ", \n ",sep = "")
    cat("with", ceiling(object$events), "events (assuming an outcome prevalence =", object$prevalence, ") and an EPP =", object$EPP, " \n \n ")

  }
}
